Advanced Capabilities week 2 day 4 update (#43)
microsoft/win-student-devs: docs/30DaysOfPWA/advanced-capabilities/04.md

    height: env(titlebar-area-height, 40px);
}
```

The `titlebar-area-x` variable gives us the distance from the left of the viewport to where the titlebar appears, with a fallback integer. The `titlebar-area-y` variable gives us the distance from the top. And `titlebar-area-width` and `titlebar-area-height` set the width and height respectively, with fallback values. The fallback values are used when the app isn't installed, when Window Controls Overlay isn't supported or the user has opted-out of the feature.

### Keeping the title bar draggable
Once you’ve positioned and styled your title bar, there’s one more step left. The window is no longer draggable if you’ve turned on window control overlay, so we need to fix this with the `app-region` CSS property. 

```
.titleBar {
    position: fixed;
    left: env(titlebar-area-x, 0);
    top: env(titlebar-area-y, 0);
    width: env(titlebar-area-width, 100%);
    height: env(titlebar-area-height, 40px);
    -webkit-app-region: drag;
    app-region: drag;
}
```

This CSS property is also experimental, so `-webkit-app-region` is needed for it to work and is only supported in Chromium-based browsers. 

## Resources 

> [MDN `display-mode`](https://aka.ms/learn-PWA/30Days-2.4/developer.mozilla.org/en-US/docs/Web/CSS/display-mode)
> [App Design: Display Modes](https://aka.ms/learn-PWA/30Days-2.4/web.dev/learn/pwa/app-design#display-modes)
> [WCO: Breaking Out of the Box](https://aka.ms/learn-PWA/30Days-2.4/alistapart.com/article/breaking-out-of-the-box)

Reviewer: ```suggestion
[MDN `display-mode`](https://aka.ms/learn-PWA/30Days-2.4/developer.mozilla.org/en-US/docs/Web/CSS/display-mode)

[App Design: Display Modes](https://aka.ms/learn-PWA/30Days-2.4/web.dev/learn/pwa/app-design#display-modes)

[WCO: Breaking Out of the Box](https://aka.ms/learn-PWA/30Days-2.4/alistapart.com/article/breaking-out-of-the-box)
```

these are showing on same line

# 2.4: Displaying content more like an app

**Author: Stephanie Stimac [@seaotta](https://twitter.com/seaotta)**

![Image of title and author.](_media/day-04.jpg)

Welcome to **Week 2 Day 4** of the [30 Days of PWA](<https://aka.ms/learn-pwa/30Days-blog>) series! Today’s blog post will walk you through the different display modes you can set for your website, and the experimental Window Controls Overlay feature which allows the customization of the title bar of your PWA.

## Display Modes 

> There are four different display modes available to choose from and each provides a different browser UI experience that you’ll want to consider when deciding which display mode is right for your progressive web application. 

Reviewer 2: I feel like we need to say that these only apply when the PWA is installed somewhere in this sentence. Proposal:

```suggestion
The way Progressive Web Apps get displayed by an operating system after installation may be different than the way the app appears when opened in a web browser. On a mobile device for example, an installed PWA can be displayed full screen, while on a desktop operating system, it can be displayed in a standalone window.

There are four different display modes available to choose from and each provides a different browser UI experience that you’ll want to consider when deciding which display mode is right for your PWA. 
```

> You set your preferred display mode via the web app manifest’s `display` property with one of the following values: "fullscreen," "standalone," "minimal-ui," or "browser":

Reviewer 2: I think we tend to use the word `member` instead of `property` when talking about root-level properties of the manifest.

}
```

In the event a browser doesn’t support a display mode, which is possible since they’re not required or support all the display mode options, the display mode will [fall back to a supported one](https://www.w3.org/TR/appmanifest/#dfn-fallback-chain) in this order: 

> fullscreen > standalone > minimal-ui > browser

### Standalone display mode 
The "standalone" display mode makes your PWA look and behave the most like a native application. It opens in a different window from the browser and hides all the browser UI elements like the address bar. It keeps standard system UI elements like a back button or a close window button. In this mode, your application can also have its own icon in the application launcher.  If standalone isn’t available, it will fall back to the "minimal-ui" display mode. Twitter uses "standalone" for their PWA.

<!-- ![Twitter displayed in its own window on Android](_media/display-standalone.png) -->
<img src="/30DaysOfPWA/advanced-capabilities/_media/display-standalone.png" alt="Twitter displayed in its own window on Android" style="max-width: 600px; display: block; margin-left: auto; margin-right: auto;" />

### Fullscreen display mode

> The "fullscreen" display mode takes up the entirety of the display area available and hides all of the browser UI elements. If "fullscreen" isn’t available, it will fall back to "standalone." Here's an example of a game that uses the "fullscreen" display mode.

Reviewer 2: Can we link to that particular game so people can install it and see what `fullscreen` does for themselves?

<img src="/30DaysOfPWA/advanced-capabilities/_media/display-fullscreen.png" alt="This PacMan game is a PWA and was built with the fullscreen display mode" style="max-width: 600px; display: block; margin-left: auto; margin-right: auto;" />

### Minimal-UI display mode 
The "minimal-ui" display mode gives your PWA a similar experience to the standalone display mode. It opens in its own window, but the application retains a minimal set of browser UI controls. The UI that is retained will vary between browsers. The OneDrive PWA uses the "minimal-ui" display mode.

![The OneDrive PWA utilizes the minimal-ui display mode](_media/display-minimal-ui.png) 

### Browser display mode
The browser display mode retains the browser experience with all the browser UI and your web app will not be installable. It will open in a normal browser window or tab. If Twitter were not using standalone mode, it would display in the browser like a normal tab. 

![Twitter in a browser window.](_media/display-desktop.png)

### Targeting different display modes with the `display-mode` media feature

> Depending on the display mode you choose for your PWA there may be elements you want to hide when the web app is open in the browser tab, especially if you’re launching in a mode like standalone or fullscreen that hides the browser UI button. You need to ensure there’s a way for people to go back if there’s no back button. Or maybe you want to modify your site styles slightly to display better when the app is in fullscreen mode. 

Reviewer 2: ```suggestion
Depending on the display mode you choose for your PWA there may be elements you want to hide when the web app is open in the browser tab, especially if you’re launching in a mode like standalone or fullscreen that hides the browser UI buttons. You need to ensure there’s a way for people to go back if there’s no back button. Or maybe you want to modify your site styles slightly to display better when the app is in fullscreen mode. 
```

You can do this by targeting certain display modes with the `display-mode` media feature. 

```
.app-button {

    display: none;

}

@media (display-mode: standalone) {

.app-button {

    display: block;

    }

> }

Reviewer 2: Looking at other code snippets in the repo, I don't think we have a convention around adding empty lines between lines of code. So it might be better to remove the empty lines here (and in other code snippets in this article).

```

> In this example, the button with the class .app-button would only be shown when the PWA is in standalone mode. You can go even further to combine media queries to change how things are displayed on smaller screens versus a desktop when a user in standalone mode to tailor the experience.

Reviewer 2: ```suggestion
In this example, the button with the class `.app-button` would only be shown when the PWA is in standalone mode. You can go even further and combine media queries to change how things are displayed on smaller screens versus a desktop when a user in standalone mode to tailor the experience.
```

## Customizing display options further with Window Controls Overlay 

> The previous display modes I mentioned currently only offer customization of the display below the title bar when in standalone mode. Window Controls Overlay is an experimental display option that will allow you to customize the title bar of your PWA with CSS and JavaScript. It provides access to the space next to the controls that minimize, maximize, or close the app’s window. 

Reviewer 2: Somewhere here, we need to say that this is only for desktop-installed PWAs so that the `below the title bar` section makes sense to readers.

### Using Window Controls Overlay 
In order to use this feature we need to add “display_override” to our web app manifest file. 

```
{

“display_override”: [“window-controls-overlay”]

}
```

The customized title bar will only show when in a separate PWA window on a desktop operating system. After our `display_override`, we would then follow up with our preferred display mode when window controls overlay isn’t available. 

```
{

“display_override”: [“window-controls-overlay”],
"display": "standalone",

}
```

Now that we’ve stated via the manifest we would like to use window controls overlay, we need to add some CSS to position the content in our title bar area. We can do that with the following [CSS environment variables](https://developer.mozilla.org/en-US/docs/Web/CSS/env()): 

- **`titlebar-area-x`**
- **`titlebar-area-y`**
- **`titlebar-area-width`**
- **`titlebar-area-height`**

Using these environment variables would look something like this: 

```
.titleBar {
    position: fixed;
    left: env(titlebar-area-x, 0);
    top: env(titlebar-area-y, 0);
    width: env(titlebar-area-width, 100%);
    height: env(titlebar-area-height, 40px);
}
```

> The `titlebar-area-x` variable gives us the distance from the left of the viewport to where the titlebar appears, with a fallback integer. The `titlebar-area-y` variable gives us the distance from the top. And titlebar-area-width and titlebar-area-height set the width and height respectively, with fallback values. 

Reviewer 2: ```suggestion
The `titlebar-area-x` variable gives us the distance from the left of the viewport to where the titlebar appears, with a fallback integer. The `titlebar-area-y` variable gives us the distance from the top. And `titlebar-area-width` and `titlebar-area-height` set the width and height respectively, with fallback values. 
```

Also worth adding some information about when the fallbacks are used (i.e. when the app is not installed, or when WCO is not supported, or when the user opted-out of WCO).

### Keeping the title bar draggable
Once you’ve positioned and styled your title bar, there’s one more step left. The window is no longer draggable if you’ve turned on window control overlay, so we need to fix this with the `app-region` CSS property. 

```
.titleBar {
    position: fixed;
    left: env(titlebar-area-x, 0);
    top: env(titlebar-area-y, 0);
    width: env(titlebar-area-width, 100%);
    height: env(titlebar-area-height, 40px);
    -webkit-app-region: drag;
    app-region: drag;
}
```

This CSS property is also experimental, so `-webkit-app-region` is needed for it to work and is only supported in Chromium-based browsers. 

## Resources 
[MDN `display-mode`](https://aka.ms/learn-PWA/30Days-2.4/developer.mozilla.org/en-US/docs/Web/CSS/display-mode)
[App Design: Display Modes](https://aka.ms/learn-PWA/30Days-2.4/web.dev/learn/pwa/app-design#display-modes)
[WCO: Breaking Out of the Box](https://aka.ms/learn-PWA/30Days-2.4/alistapart.com/article/breaking-out-of-the-box)

## Summary

> Display modes are an easy way to make your PWA feel like a native app and are set via the web app manifest `display` property. The ability to further customize the title bar area for an even greater native app feel when your PWA is displayed in a separate window is available via the experimental window control overlay setting available in Chrome and Edge. 

Reviewer 2: ```suggestion
Display modes are an easy way to make your PWA feel like a native app and are set via the web app manifest `display` member. The ability to further customize the title bar area for an even greater native desktop app feel when your PWA is displayed in a separate window is available via the experimental Window Controls Overlay setting available in Chromium browsers. 
```

<!-- ![Twitter displayed in its own window on Android](_media/display-standalone.png) -->
<img src="/30DaysOfPWA/advanced-capabilities/_media/display-standalone.png" alt="Twitter displayed in its own window on Android" style="max-width: 600px; display: block; margin-left: auto; margin-right: auto;" />

### Fullscreen display mode
The "fullscreen" display mode takes up the entirety of the display area available and hides all of the browser UI elements. If "fullscreen" isn’t available, it will fall back to "standalone." Here's an example of a game that uses the "fullscreen" display mode.

<!-- ![This PacMan game is a PWA and was built with the fullscreen display mode](_media/display-fullscreen.png) -->

> <img src="/30DaysOfPWA/advanced-capabilities/_media/display-fullscreen.png" alt="This PacMan game is a PWA and was built with the fullscreen display mode" style="max-width: 600px; display: block; margin-left: auto; margin-right: auto;" />

Reviewer: ```suggestion
<img src="_media/display-fullscreen.png" alt="This PacMan game is a PWA and was built with the fullscreen display mode" style="max-width: 600px; display: block; margin-left: auto; margin-right: auto;" />
```

First two images aren't displaying.
